Add readyz endpoint to apiserver, modify poststarthooks health checking behavior

logicalhan opened the pull request:
This PR adds a readyz endpoint to apiserver so that we can export readiness information. It also modifies poststarthooks to do two things: 

1. liveness health checking returns true until an explicit max startup sequence period has elapsed (the time one expects apiserver boot-sequencing to have completed) .
2. readiness health checking should immediately return whether that poststarthook has completed (this is the current behavior of our post start hook health checks).

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Adds a proper readiness endpoint to the kube-apiserver.

**Which issue(s) this PR fixes**:

Fixes #55453

```release-note
The kube-apiserver has improved behavior for both startup and shutdown sequences and also now exposes `\readyz` for readiness checking. Readyz includes all existing healthz checks but also adds a shutdown check. When a cluster admin initiates a shutdown, the kube-apiserver will try to process existing requests (for the duration of request timeout) before killing the apiserver process.  

The apiserver also now takes an optional flag "--maximum-startup-sequence-duration". This allows you to explicitly define an upper bound on the apiserver startup sequences before healthz begins to fail. By keeping the kubelet liveness initial delay short, this can enable quick kubelet recovery as soon as we have a boot sequence which has not completed in our expected time frame, despite lack of completion from longer boot sequences (like RBAC). Kube-apiserver behavior when the value of this flag is zero is backwards compatible (this is as the defaulted value of the flag).
```

/sig api-machinery

logicalhan commented:
/assign @sttts

sttts commented:
First commit message needs a little rework https://chris.beams.io/posts/git-commit/

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -469,7 +473,8 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 		disabledPostStartHooks: c.DisabledPostStartHooks,
 
-		healthzChecks: c.HealthzChecks,
+		livenessHealthChecks:  c.HealthzChecks,
```
healthChecks

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -469,7 +473,8 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 
-		healthzChecks: c.HealthzChecks,
+		livenessHealthChecks:  c.HealthzChecks,
+		readinessHealthChecks: c.ReadyzChecks,
```
readinessChecks

Would not mix both terms.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -145,9 +145,11 @@ type GenericAPIServer struct {
-	healthzLock    sync.Mutex
-	healthzChecks  []healthz.HealthzChecker
-	healthzCreated bool
+	healthzLock                 sync.Mutex
```
checksLock

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -306,6 +308,7 @@ func (s preparedGenericAPIServer) NonBlockingRun(stopCh <-chan struct{}) error {
 		}
 	}
 
+	s.installReadyz(stopCh)
```
this is strange to install it here. I my PR I had another stop check, didn't I?

logicalhan replied:
No, it was like this, but I will plumb the stopCh logic as we discussed offline.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,73 @@ package server
 
-	if s.healthzCreated {
+	if s.livenessHealthChecksLocked || s.readinessHealthChecksLocked {
+		return fmt.Errorf("unable to add because the healthz endpoint has already been created")
```
message does not match both cases

sttts replied:
Still open

logicalhan replied:
It is fixed.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:
```diff
@@ -116,6 +124,7 @@ func InstallPathHandler(mux mux, path string, checks ...HealthzChecker) {
 	klog.V(5).Info("Installing healthz checkers:", formatQuoted(checkerNames(checks...)...))
 
 	mux.Handle(path, handleRootHealthz(checks...))
+	// TODO: let's remove this in a future version
```
then we have to deprecate it, in code, the release notes and the docs.

logicalhan replied:
I'll leave this for the next release :)

sttts commented on staging/src/k8s.io/apiserver/pkg/server/hooks.go:
```diff
@@ -19,14 +19,15 @@ package server
 import (
 	"errors"
 	"fmt"
+	"k8s.io/apimachinery/pkg/util/clock"
```
order

sttts commented on staging/src/k8s.io/apiserver/pkg/server/hooks.go:
```diff
@@ -210,28 +216,65 @@ func runPreShutdownHook(name string, entry preShutdownHookEntry) error {
-// postStartHookHealthz implements a healthz check for poststarthooks.  It will return a "hookNotFinished"
+type postStartHookHealthChecks struct {
+	livenessCheck  delayedLivenessCheck
+	readinessCheck postStartReadinessCheck
```
Why is this necessary?

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -145,9 +145,12 @@ type GenericAPIServer struct {
+	livenessHealthChecksLocked  bool
+	readinessHealthChecks       []healthz.HealthzChecker
+	readinessHealthChecksLocked bool
+	postStartHookMaxDurationInSecs    int64
```
time.Duration

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,117 @@ package server
-
 	"k8s.io/apiserver/pkg/server/healthz"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"net/http"
```
order, golang import first, separated

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,117 @@ package server
+}
+
+// AddReadyzChecks allows you to add a HealthzCheck to readyz.
+func (s *GenericAPIServer) AddReadyzChecks(checks ...healthz.HealthzChecker) error {
```
if we use "health check" as the general term, better drop the `z` everywhere where don't mean the actual `/healthz` endpoint

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,117 @@ package server
 
-	s.healthzChecks = append(s.healthzChecks, checks...)
+	if s.readinessHealthChecksLocked {
+		return fmt.Errorf("unable to add because the healthz endpoint has already been created")
```
readyz

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -251,21 +262,24 @@ type AuthorizationInfo struct {
+		MaxMutatingRequestsInFlight:            200,
+		RequestTimeout:                         time.Duration(60) * time.Second,
+		MinRequestTimeout:                      1800,
+		PostStartHookHealthzGracePeriod: time.Duration(0),
```
is this correct formatting? looks like wrong indention

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -469,7 +482,10 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
-		healthzChecks: c.HealthzChecks,
+		healthChecks:             c.HealthzChecks,
+		readinessChecks:          c.ReadyzChecks,
+		readinessStopCh:          make(chan struct{}),
```
what does it do, document with a oneliner

logicalhan replied:
sure

sttts replied:
actually below in the type

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -145,9 +145,13 @@ type GenericAPIServer struct {
-	healthzLock    sync.Mutex
-	healthzChecks  []healthz.HealthzChecker
-	healthzCreated bool
+	healthzLock              sync.Mutex
```
healthzReadyzLock, better split it into two.

logicalhan replied:
sure.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -145,9 +145,13 @@ type GenericAPIServer struct {
+	readinessChecks          []healthz.HealthzChecker
+	readinessChecksLocked    bool
+	postStartHookGracePeriod time.Duration
+	readinessStopCh          chan struct{}
```
what does it do, document with a oneliner

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,117 @@ package server
-	if s.healthzCreated {
-		return fmt.Errorf("unable to add because the healthz endpoint has already been created")
+	if s.healthChecksLocked || s.readinessChecksLocked {
+		return fmt.Errorf("unable to add because the healthz or readyz endpoint has already been created")
```
make it two if blocks with two messages

logicalhan replied:
Actually I refactored it away.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,117 @@ package server
+// AddHealthzCheck adds HealthzCheck(s) to both healthz and readyz. All healthz checks
+// are automatically added to readyz, since we want to avoid the situation where the
+// apiserver is ready but not live.
+func (s *GenericAPIServer) AddDelayedHealthzCheck(check healthz.HealthzChecker) error {
```
add a delay parameter and call this with delay=0 from AddHealthzCheck to unify the code. Also move them next to each other to make it easier to maintain.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,117 @@ package server
+	return nil
+}
+
+func delayedHealthCheck(check healthz.HealthzChecker, clock clock.Clock, delay time.Duration) healthz.HealthzChecker {
```
:+1:

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz_test.go:
```diff
@@ -0,0 +1,75 @@
+import (
+"k8s.io/apimachinery/pkg/util/clock"
+"testing"
+"time"
```
indention

sttts commented on staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go:
```diff
@@ -185,6 +192,11 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
 		"--min-request-timeout for specific types of requests.")
 
+	fs.DurationVar(&s.PostStartHookGracePeriod, "post-start-hook-grace-period", s.RequestTimeout, ""+
+		"An optional field indicating the duration that poststarthook liveness health checks will return true "+
```
Better avoid liveness check as a term and use `/healthz` instead.

sttts replied:
Would not use poststarthook as a term. Better something like `Period after which /healthz returns failure if the apiserver does not become ready (reported via /readyz) on startup. Zero for immediate /healthz failure until it is ready.`.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -208,6 +208,7 @@ func (c completedConfig) NewWithDelegate(delegationTarget genericapiserver.Deleg
 	s.GenericAPIServer.AddPostStartHookOrDie("apiservice-status-available-controller", func(context genericapiserver.PostStartHookContext) error {
 		// if we end up blocking for long periods of time, we may need to increase threadiness.
 		go availableController.Run(5, context.StopCh)
+		time.Sleep(time.Duration(30) * time.Second)
```
debugging code

logicalhan replied:
removed :)

logicalhan commented:
/test pull-kubernetes-integration

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -155,6 +157,15 @@ type Config struct {
+	// is obscure. Specifically, if the poststarthook must complete by the following timeout:
+	// the static pod's configured InitialDelaySeconds + (FailureThreshold * PeriodSeconds)
+	// If none of these values are overridden, this means that if the kube-apiserver boot-sequence does not complete
+	// in 45 seconds, the kubelet will restart the apiserver.
```
those 45 seconds depend on the context. Also not every kube-apiserver runs in a static pod.

sttts replied:
make this shorter and more crisp. Background is nice, but I would remove the whole paragraph about the static pod environment. It's not generic.

logicalhan replied:
Ok.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -145,9 +145,16 @@ type GenericAPIServer struct {
-	healthzChecks  []healthz.HealthzChecker
-	healthzCreated bool
+	healthzLock              sync.Mutex
+	healthChecks             []healthz.HealthzChecker
```
healthzChecks

logicalhan replied:
sure

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -145,9 +145,16 @@ type GenericAPIServer struct {
+	healthChecks             []healthz.HealthzChecker
+	healthChecksLocked       bool
+	readyzLock               sync.Mutex
+	readinessChecks          []healthz.HealthzChecker
```
readyzChecks, for consistency in naming

logicalhan replied:
okay

sttts replied:
not done yet?

logicalhan replied:
fixed

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +85,43 @@ func TestRun(t *testing.T) {
+func TestReadyz(t *testing.T) {
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--post-start-hook-grace-period", "10s"}, framework.SharedEtcd())
+	defer server.TearDownFn()
+	defer time.Sleep(10 * time.Second)
```
why that sleep?

logicalhan replied:
It's removed. There were some clean-up issues in previous iterations of this test.

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +85,43 @@ func TestRun(t *testing.T) {
 }
 
+func TestReadyz(t *testing.T) {
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--post-start-hook-grace-period", "10s"}, framework.SharedEtcd())
```
For better testing you want a `DontWaitForHealthz` in `TestServerInstanceOptions`.

logicalhan replied:
When the grace period is enabled, healthz returns true immediately, so we actually don't need to explicitly configure that.

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +85,43 @@ func TestRun(t *testing.T) {
+	defer server.TearDownFn()
+	defer time.Sleep(10 * time.Second)
+	t.Run("test post start hook grace period", func(t *testing.T) {
+		if server.ServerOpts.GenericServerRunOptions.PostStartHookGracePeriod.Seconds() != 10 {
```
the options are directly mapped to flags. There is not much value of this test I fear.

logicalhan replied:
Revised.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -134,6 +134,8 @@ type Config struct {
 	DiscoveryAddresses discovery.Addresses
 	// The default set of healthz checks. There might be more added via AddHealthzChecks dynamically.
 	HealthzChecks []healthz.HealthzChecker
+	// The default set of readyz-only checks. There might be more added via AddReadyzChecks dynamically.
```
this is the config struct, not really the default, but the actual set of checks.

logicalhan replied:
ah, this was a stray comment from an iteration long ago.

logicalhan replied:
It turns out I actually stole this comment from you. You wrote the initial comment two lines above for HealthzChecks three years ago.

sttts replied:
Since that time it was moved around 7 times probably. Never trust existing code or comments ;-)

logicalhan replied:
It actually does make sense. We do assign the default checks to this field. And additional checks are, in fact, added dynamically.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go:
```diff
@@ -185,6 +192,11 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
 		"it out. This is the default request timeout for requests but may be overridden by flags such as "+
 		"--min-request-timeout for specific types of requests.")
 
+	fs.DurationVar(&s.PostStartHookGracePeriod, "post-start-hook-grace-period", s.PostStartHookGracePeriod, ""+
```
or healthz-grace-period?

@lavalamp you like clear flag names. Is that good ^ ?

logicalhan replied:
I'll defer this one to @lavalamp

lavalamp replied:
How about "--maximum-startup-duration"? --maximum-post-start-hook-duration? --maximum-startup-sequence-duration?

logicalhan replied:
Opted for `--maximum-startup-sequence-duration` as it most closely aligns with the flag introduced in #74416.

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +85,83 @@ func TestRun(t *testing.T) {
+}
+
+func TestDelayedHealthzAndReadyz(t *testing.T) {
+	hc := delayedCheck{time.Now().Add(10 * time.Second)}
```
why not make that synchronously controlled? I.e. put a protected failing bool inside and then set that during startup and later after startup.

sttts replied:
Also a ten second runtime of a test is not good. This should be as fast as possible. Use `wait.ForeverTestTimeout` as upper limit everywhere, but otherwise it should run without fixed duration values.

logicalhan replied:
sure

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +85,83 @@ func TestRun(t *testing.T) {
+		t.Fatalf("Readyz should start unhealthy")
+	}
+	errCh := make(chan error)
+	err = wait.Poll(100*time.Millisecond, 10*time.Second, func() (bool, error) {
```
hardcoded times are suspicious

logicalhan replied:
we need to wait longer than the grace period. i will get rid of the other polling instances though.

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,82 @@ func TestRun(t *testing.T) {
 	}
 }
 
```
golang‘ism: move helpers to the bottom

logicalhan replied:
👍

logicalhan commented:
/test pull-kubernetes-integration

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,82 @@ func TestRun(t *testing.T) {
+	readyzRes.StatusCode(&readyzStatus)
+	if readyzStatus == 200 {
+		t.Fatalf("Readyz should start unhealthy")
+	}
```
how about a helper to condense those repeated 7 lines to one?

logicalhan replied:
so funny, i actually had this last night but it was lost in one of my changes.

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,82 @@ func TestRun(t *testing.T) {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	readyzRes := client.CoreV1().RESTClient().Get().AbsPath("/readyz").Do()
+	readyzStatus := 0
```
var readyzStatus int

logicalhan replied:
done

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,82 @@ func TestRun(t *testing.T) {
+	readyzStatus := 0
+	readyzRes.StatusCode(&readyzStatus)
+	if readyzStatus == 200 {
+		t.Fatalf("Readyz should start unhealthy")
```
unready

logicalhan replied:
fixed

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,82 @@ func TestRun(t *testing.T) {
+	readyzRes := client.CoreV1().RESTClient().Get().AbsPath("/readyz").Do()
+	readyzStatus := 0
+	readyzRes.StatusCode(&readyzStatus)
+	if readyzStatus == 200 {
```
http.StatusOK

logicalhan replied:
done

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,82 @@ func TestRun(t *testing.T) {
+	if readyzStatus == 200 {
+		t.Fatalf("Readyz should start unhealthy")
+	}
+	errCh := make(chan error)
```
why a channel?

logicalhan replied:
copied from above. it is unnecessary.

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,82 @@ func TestRun(t *testing.T) {
+	}
+	readyzRes := client.CoreV1().RESTClient().Get().AbsPath("/readyz").Do()
+	readyzStatus := 0
+	readyzRes.StatusCode(&readyzStatus)
```
what about healthz?

logicalhan replied:
It is checked immediately above it in the polling loop.

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,82 @@ func TestRun(t *testing.T) {
+	readyzStatus = 0
+	readyzRes.StatusCode(&readyzStatus)
+	if readyzStatus != 200 {
+		t.Fatalf("Readyz should be healthy")
```
nit: lowercase

logicalhan replied:
fixed

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,117 @@ package server
+	defer s.healthzLock.Unlock()
+	c := clock.RealClock{}
+	if s.healthzChecksLocked {
+		return fmt.Errorf("unable to add because either the healthz endpoint has already been created")
```
-either

logicalhan replied:
removed

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,117 @@ package server
+		s.healthzChecks = append(s.healthzChecks, delayedHealthCheck(check, c, s.postStartHookGracePeriod))
+	}
+
+	err := s.AddReadyzChecks(checks...)
```
return s.AddReadyzChecks(checks...)

logicalhan replied:
nice

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:
```diff
@@ -102,6 +102,14 @@ func InstallHandler(mux mux, checks ...HealthzChecker) {
 }
 
+// InstallHandler registers handlers for health checking on the path
+// "/healthz" to mux. *All handlers* for mux must be specified in
```
update doc

logicalhan replied:
fixed

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,55 @@ func TestRun(t *testing.T) {
+func endpointReturnsStatusOK(client *kubernetes.Clientset, path string) bool {
+	var status int
+	res := client.CoreV1().RESTClient().Get().AbsPath(path).Do()
+	status = 0
```
is this necessary?

logicalhan replied:
Is there another way to derive the status code?

sttts replied:
The way is fine. Was just questioning the assignment of zero. It's already done by the declaration.

logicalhan replied:
oh duh. yeah sure, i can remove it.

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,55 @@ func TestRun(t *testing.T) {
 }
 
+func endpointReturnsStatusOK(client *kubernetes.Clientset, path string) bool {
+	var status int
```
golang'ism move delcaration before first use

logicalhan replied:
sure

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,55 @@ func TestRun(t *testing.T) {
+	}
+	// we need to wait longer than our grace period
+	err = wait.Poll(100*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
+		if !endpointReturnsStatusOK(client, "/healthz") {
```
return !endpointReturnsStatusOK(client, "/healthz"), nil

logicalhan replied:
like it :)

sttts commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,55 @@ func TestRun(t *testing.T) {
+	}
+	hc.makeHealthy()
+	err = wait.Poll(100*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
+		if endpointReturnsStatusOK(client, "/healthz") {
```
return endpointReturnsStatusOK(client, "/healthz"), nil

logicalhan replied:
oh, that's nice :)

sttts commented:
about the release note:
* what does "effectively" mean?
* don't talk about the kubelet. Using a static pod is just one special case. Be explicit about the actual endpoints.
* what is a graceful shutdown health check (I know, but other people won't understand) ? Be explicit about what happens with /healthz and /readyz on shutdown.

sttts commented:
Sgtm. Tests still are not happy.

sttts commented:
Please squash.

Commit:
add readyz endpoint for kube-apiserver readiness checks

add startup sequence duration and readyz endpoint

add rbac bootstrapping policy for readyz

add integration test around grace period and readyz

rename startup sequence duration flag

copy health checks to fields

rename health-check installed boolean, refactor clock injection logic

cleanup clock injection code

remove todo about poststarthook url registration from healthz

sttts commented:
/retest

sttts commented:
/lgtm

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/options/server_run_options.go:
```diff
@@ -185,6 +192,11 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
 		"--min-request-timeout for specific types of requests.")
 
+	fs.DurationVar(&s.PostStartHookGracePeriod, "post-start-hook-grace-period", s.PostStartHookGracePeriod, ""+
+		"An optional field indicating the period after which /healthz returns failure if the apiserver does not become "+
```
This option represents the maximum amount of time it should take for apiserver to complete its startup sequence and become healthy. From apiserver's start time to when this amount of time has elapsed, /healthz will assume that unfinished post-start hooks will complete successfully and therefore return true.

logicalhan replied:
I am going to steal this word for word.

sttts commented:
/retest

no logs

sttts commented:
/lgtm

sttts commented:
/lgtm

logicalhan commented:
/test pull-kubernetes-integration

logicalhan commented:
/assign @lavalamp

logicalhan commented:
/test pull-kubernetes-e2e-gce-rbe

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -251,13 +259,15 @@ type AuthorizationInfo struct {
 		DisabledPostStartHooks:      sets.NewString(),
-		HealthzChecks:               []healthz.HealthzChecker{healthz.PingHealthz, healthz.LogHealthz},
+		HealthzChecks:               defaultHealthChecks,
+		ReadyzChecks:                defaultHealthChecks,
```
if you expect people to modify these lists, you probably want two different lists, otherwise append won't do what you expect.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -145,9 +145,16 @@ type GenericAPIServer struct {
-	healthzCreated bool
+	healthzLock                sync.Mutex
+	healthzChecks              []healthz.HealthzChecker
+	healthzChecksLocked        bool
```
"Locked" makes it sound like it's got to do with the mutex, I suggest "Installed" or "Finalized"?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,113 @@ package server
+// grace period has not yet elapsed. One may want to set a grace period in order to prevent
+// the kubelet from restarting the kube-apiserver due to long-ish boot sequences. Readyz health
+// checks have no grace period, since we want readyz to fail while boot has not completed.
+func (s *GenericAPIServer) AddDelayedHealthzChecks(delay time.Duration, checks ...healthz.HealthzChecker) error {
```
Do we expect this to be called multiple times? It seems a little odd for the clock to be created each time this is called. Maybe not important to fix.

logicalhan replied:
Yeah, I will admit that I thought about this but didn't do anything about it. I'll move it up into the GenericAPIServer object.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:
```diff
@@ -116,6 +124,7 @@ func InstallPathHandler(mux mux, path string, checks ...HealthzChecker) {
 	klog.V(5).Info("Installing healthz checkers:", formatQuoted(checkerNames(checks...)...))
 
 	mux.Handle(path, handleRootHealthz(checks...))
+	// TODO: let's remove this in a future version
```
It'd be good to (in a followup) state the reasoning--TODOs are sometimes blindly followed.

logicalhan replied:
Sure thing.

lavalamp commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -85,6 +86,48 @@ func TestRun(t *testing.T) {
+		t.Fatalf("readyz should start unready")
+	}
+	// we need to wait longer than our grace period
+	err = wait.Poll(100*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
```
but... but... you used a clock earlier?

logicalhan replied:
Yeah, I can rewrite this to leverage the clock.

logicalhan replied:
I am going to avoid doing this, because I can't think of a good way to plumb the clock through the  integration testserver.

sttts replied:
no clock magic here. This is an (e2e-like) integration test. We want the normal timing behaviour.

logicalhan replied:
👍

lavalamp commented:
I'm still a little nervous about releasing this without an alpha warning or something.

deads2k commented:
I like getting readiness, but I'm inclined to wait until 1.16 to merge this.

I think some explanation of how the flag for ignoring incomplete post-starthooks for a while is better than setting a longer timeout on a health check may be worthwhile too.  I'm not strictly against the idea, but I'm not completely clear on why I'd choose specifying a flag over specifying a health check timeout.

dstrebel commented:
Will this be moving to v1.16?

logicalhan commented:
> Will this be moving to v1.16?

Yes.

logicalhan commented:
/test pull-kubernetes-e2e-gce-100-performance

sttts commented on staging/src/k8s.io/apiserver/pkg/server/healthz.go:
```diff
@@ -18,28 +18,112 @@ package server
+// grace period has not yet elapsed. One may want to set a grace period in order to prevent
+// the kubelet from restarting the kube-apiserver due to long-ish boot sequences. Readyz health
+// checks have no grace period, since we want readyz to fail while boot has not completed.
+func (s *GenericAPIServer) AddDelayedHealthzChecks(delay time.Duration, c clock.Clock, checks ...healthz.HealthzChecker) error {
```
exposing clock here is wrong. We have it in s already. No need to have an extra clock here.

logicalhan replied:
👍

sttts commented:
/lgtm

logicalhan commented:
/assign @deads2k

logicalhan commented:
> I like getting readiness, but I'm inclined to wait until 1.16 to merge this.
> 
> I think some explanation of how the flag for ignoring incomplete post-starthooks for a while is better than setting a longer timeout on a health check may be worthwhile too. I'm not strictly against the idea, but I'm not completely clear on why I'd choose specifying a flag over specifying a health check timeout.

@deads2k I've updated the release notes as discussed, amended with some of @sttts' suggestions (thanks @sttts!)

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -145,9 +146,17 @@ type GenericAPIServer struct {
-	healthzLock    sync.Mutex
-	healthzChecks  []healthz.HealthzChecker
-	healthzCreated bool
+	healthzLock                sync.Mutex
```
optional nit: we're up to five fields for health checks at this point, three are the same as readiness.  I would consider a struct.

logicalhan replied:
That sounds like a good idea. I'll refactor this in a follow-up PR.

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go:
```diff
@@ -107,6 +115,13 @@ func InstallPathHandler(mux mux, path string, checks ...HealthzChecker) {
 	klog.V(5).Info("Installing healthz checkers:", formatQuoted(checkerNames(checks...)...))
 
 	mux.Handle(path, handleRootHealthz(checks...))
+	// TODO: Consider removing this in a future version. This basically adds a bunch
```
I'm ok with hiding them.  I'm not ok with removing the ability to get detailed information about each health check.

logicalhan replied:
As discussed offline, I'll remove the TODO as it is an orthogonal change.

deads2k commented:
Spoke on slack.  We'll remove the TODO so that remote debugging with proper permissions is still possible.

/approve

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -493,12 +504,16 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
-		healthzChecks: c.HealthzChecks,
+		healthzChecks:              c.HealthzChecks,
+		readyzChecks:               c.ReadyzChecks,
+		readinessStopCh:            make(chan struct{}),
```
does the wiring really need these 3 in the config or would GenericAPIServer be enough?

logicalhan replied:
I'm not sure I follow what you are saying.. the portion of the code you are commenting on is in fact the GenericAPIServer. Specifically this is where we initialize the GenericAPIServer object.

sttts commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -493,12 +504,16 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 
 		enableAPIResponseCompression: c.EnableAPIResponseCompression,
 		maxRequestBodyBytes:          c.MaxRequestBodyBytes,
+		healthzClock:                 clock.RealClock{},
```
does the wiring really need this in the config?

logicalhan replied:
My thought was that it would be wise to initialize the GenericAPIServer with the clock since we can call AddPostStartHook arbitrarily on the GenericAPIServer object (prior to the installHealthz and installReadyz calls, in fact).

soggiest commented:
/milestone v1.16

sttts commented:
/lgtm

tedyu commented on test/integration/master/kube_apiserver_test.go:
```diff
@@ -253,3 +296,27 @@ func TestReconcilerMasterLeaseMultiMoreMasters(t *testing.T) {
+
+func (h *delayedCheck) Check(req *http.Request) error {
+	h.healthLock.Lock()
+	defer h.healthLock.Unlock()
```
If healthLock is RWMutex, we can take RLock in this func, making the check more efficient.
